Implement WaitForExitAsync for System.Diagnostics.Process

MattKotsenas opened the pull request:
This PR implements the API proposal dotnet/corefx#34689: Add a task-based `WaitForExitAsync()` for the `Process` class.

In order to ensure that the `Exited` event is never lost, it is the caller's responsibility to ensure that `EnableRaisingEvents` is `true` _prior_ to calling `WaitForExitAsync()`. A new `InvalidOperationException` with an appropriate error message is introduced to enforce those semantics.

Additionally, per feedback, `WaitForExitAsync` returns a `Task` instead of more closely matching the sync version and returning `Task<bool>`. To determine if the process has exited,
callers can check `HasExited` after the await (also cancellation follows the async pattern of throwing an `OperationCanceledException` and setting canceled on the task).

Unit tests for `WaitForExit()` were duplicated to add async versions, and one sync test was missing an assert, so I added it.

MattKotsenas committed:
Add Process.WaitForExitAsync() and associated unit tests

Add a task-based `WaitForExitAsync()` for the `Process` class, and
duplicate existing `WaitForExit()` tests to add async versions.

In order to ensure that the `Exited` event is never lost, it is the
callers responsibility to ensure that `EnableRaisingEvents` is `true`
_prior_ to calling `WaitForExitAsync()`. A new
`InvalidOperationException` is introduced to enforce those semantics.

MattKotsenas committed:
Review feedback: Change WaitForExitAsync to return Task

Per review feedback, change `WaitForExitAsync` to return a `Task`
instead of `Task<bool>`. Now, to determine if the process has exited,
callers should check `HasExited` after the await, and cancellation
follows the async pattern of setting canceled on the task.

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -24,6 +24,28 @@ public void MultipleProcesses_StartAllKillAllWaitAll()
+                    var task = p.WaitForExitAsync(cts.Token);
+                    await task;
+                    Assert.True(task.IsCompletedSuccessfully);
+                    Assert.False(task.IsCanceled);
```
Is this necessary? Seems like you're testing that `await` functions correctly here, not anything specific to `WaitForExitAsync`.

MattKotsenas replied:
Since I was working with a `TaskCompletionSource` directly I figured it was better check these properties as well.

If you prefer I remove them let me know and I can do that.

scalablecory replied:
I'd prefer them removed; avoids some noise in the tests.

MattKotsenas replied:
Removed!

scalablecory reviewed (approved):
lgtm, though minor efficiency change noted. I also think the tests can be shortened a little bit.

scalablecory commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,43 @@ public bool WaitForExit(int milliseconds)
+                    return;
+                }
+
+                using (cancellationToken.Register(() => tcs.TrySetCanceled()))
```
Use the overload that lets you pass an object in here, to avoid allocating a delegate per call.

stephentoub replied:
It could also use UnsafeRegister instead of Register.  We don't need ExecutionContext to flow into the callback here, because there's no user code getting invoked. e.g.
```C#
using (cancellationToken.UnsafeRegister(s => ((TaskCompletionSource<object>)s).TrySetCanceled(), tcs))
```

Ideally, we'd also pass `cancellationToken` into `TrySetCanceled` as an argument, so that the thrown OperationCanceledException contains the relevant token.  Though to do that without incurring additional allocations would require creating a TCS-derived type that could store the CancellationToken on it as a field... and it's probably not worth it.  An alternate that's not quite as good but still relatively pay for play would be to change the await below to be:
```C#
try
{
    await tcs.Task.ConfigureAwait(false);
}
catch (OperationCanceledException)
{
    cancellationToken.ThrowIfCancellationRequested();
    throw;
}
```

scalablecory replied:
> Though to do that without incurring additional allocations would require creating a TCS-derived type that could store the CancellationToken on it as a field... and it's probably not worth it.

We could promote the `TaskCompletionSourceWithCancellation` class from System.Net.Http to common and reuse it here.

MattKotsenas replied:
I'm happy to use `TaskCompletionSourceWithCancellation`, and it'd be nice to do at least one thing "ideally" ;).

I've not introduced a common class before; from looking at the csproj, it looks like the process is to move the class into the appropriate namespace under common, then add the file as a link in both libraries? Please let me know if I'm missing anything!

scalablecory replied:
You'll want to move the file into the Common\src\System\Threading\Tasks directory, change namespace, make sure the original library still compiles.

MattKotsenas replied:
I've gone ahead and moved `TaskCompletionSourceWithCancellation` to common, wired it up to both projects, and verified locally that it still compiles and test pass in both.

scalablecory commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,43 @@ public bool WaitForExit(int milliseconds)
+        public async Task WaitForExitAsync(CancellationToken cancellationToken = default)
+        {
+            if (!Associated) { throw new InvalidOperationException(SR.NoAssociatedProcess); }
+            if (!EnableRaisingEvents) { throw new InvalidOperationException(SR.EnableRaisingEventsRequired); }
```
We probably shouldn't set it for them, but if it would "just work" instead of requiring them set it that'd be great.

stephentoub replied:
I agree this behavior of throwing is strange.  While not ideal, I think it'd better to make this set EnableRaisingEvents to true if it's false, and just document clearly that it does so (otherwise, we'll just be forcing the developer to do so themselves).  What's the downside?  Just inconsistency with the Exited property?  If we can't stomach that, then I also agree we could make it work without changing the value actually visible from EnableRaisingEvents.

MattKotsenas replied:
Initially I was worried about changing the value of EnableRaisingEvents automatically on the developer, which is why I went the exception route, but having things just work is appealing.

@scalablecory, are you opposed to Stephen's suggestion of setting EnableRaisingEvents to true?

scalablecory replied:
I'm fine setting it to true. I just realized `EnableRaisingEvents` only seems to effect the `Exited` event, not either of the others.

MattKotsenas replied:
I've updated the code to not require the user to set `EnableRaisingEvents`. Doing so introduces a chance for an exception if the process exits before setting `EnableRaisingEvents`, so I've added a comment to the top of the method explaining the set of possible cases and how they're handled.

Feedback and any thoughts are appreciated!

MihaZupan commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -54,6 +97,32 @@ public void MultipleProcesses_ParallelStartKillWait()
+        public async Task MultipleProcesses_ParallelStartKillWaitAsync()
+        {
+            const int Tasks = 4, ItersPerTask = 10;
+            Action work = async () =>
```
```suggestion
            Func<Task> work = async () =>
```

This has to be a `Func<Task>`, right now you are creating async voids and the `Task.WhenAll` will return right away.

MattKotsenas replied:
Thanks for this! I've updated the code.

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,43 @@ public bool WaitForExit(int milliseconds)
+            }
+            finally
+            {
+                Exited -= ExitedHandler;
```
This is going to allocate another delegate.  A better way to do this would be:
```C#
EventHandler handler = (s, e) => tcs.TrySetResult(null);
Exited += handler;
try
{
    ...
}
finally
{
    Exited -= handler;
}
```

MattKotsenas replied:
Thanks for this! I've updated the code.

MattKotsenas committed:
Remove asserts on Task properties

Per code review feedback, remove the asserts that verify that the Task
returned by WaitForExitAsync is completed successfully / canceled, as
they're essentially Task tests and not relevant to WaitForExitAsync.

MattKotsenas committed:
Fix unit test to not create async void

Per code review feedback, fix the
MultipleProcesses_ParallelStartKillWaitAsync test to make work a
`Func<Task>` instead of an `Action` since we're await-ing it.

MattKotsenas committed:
Remove implicit delegate creation for ExitHandler

Per code review feedback, converting ExitHandler from a local function
to an `EventHandler` to avoid the extra delegate allocation to convert
between the types.

MattKotsenas committed:
Flow CancellationToken to OperationCanceledExceptions

Per code review, register the `TaskCompletionSource` with the
`CancellationToken` so that if an `OperationCanceledException` is thrown
the relevant token is available on
`OperationCanceledException.CancellationToken`.

To accomplish that the `TaskCompletionSourceWithCancellation` class
that's internal to `System.Net.Http` is moved to Common and consumed in
both places.

Unit tests are also updated to verify that the cancellation token passed
in is the same one returned from
`OperationCanceledException.CancellationToken`.

MattKotsenas committed:
Do not require EnableRaisingEvents to already be true

Per code review feedback, update `WaitForExitAsync` to not require the
user to call `EnableRaisingEvents` first.

Setting `EnableRaisingEvents` ourselves introduces the chance for an
exception in the case where the process has already exited, so I've
added a comment to the top of the method to detail the possible paths
through the code and added comment annotations for each case.

Lastly, I updated the unit tests to remove `EnableRaisingEvents` calls.

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,94 @@ public bool WaitForExit(int milliseconds)
+                }
+
+                // CASE 1.1 & CASE 3.1: Process exits or is canceled here
+                await tcs.WaitWithCancellationAsync(cancellationToken);
```
```suggestion
                await tcs.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false);
```

MattKotsenas replied:
Thanks! Updated.

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,94 @@ public bool WaitForExit(int milliseconds)
+                // CASE 3: User already enabled events (no-op)
+                EnableRaisingEvents = true;
+            }
+            catch (InvalidOperationException)
```
EnableRaisingEvents is throwing an exception if the associated process already exited?  Can you share a repro of that?

MattKotsenas replied:
Sure! I originally hit this in the unit test `System.Diagnostics.Tests.ProcessWaitingTests.SingleProcess_CopiesShareExitAsyncInformation`.

Here's a console app that repros the issue so you don't need to be inside an xunit project:

```csharp
using System;
using System.Diagnostics;
using System.Threading;

class Program
{
    static void Main(string[] args)
    {
        var timeout = TimeSpan.FromSeconds(5);

        var p = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "cmd.exe",
                Arguments = $"/c timeout /t {timeout.TotalSeconds} /nobreak",
                UseShellExecute = false,
            }
        };

        Console.WriteLine("Staring main process...");
        p.Start();
        Console.WriteLine($"PID = {p.Id}");

        Console.WriteLine("Creating 'copy' of process by PID");
        var copy = Process.GetProcessById(p.Id);

        Console.WriteLine("Waiting for main process to exit...");
        Thread.Sleep(timeout * 2);

        Console.WriteLine("Enabling events on main process");
        p.EnableRaisingEvents = true;

        Console.WriteLine("Enabling events on 'copy'");
        copy.EnableRaisingEvents = true;

        Console.WriteLine("Done!");
    }
}
```

And here's the output:

```
Staring main process...
PID = 37344
Creating 'copy' of process by PID
Waiting for main process to exit...

Waiting for 0 seconds, press CTRL+C to quit ...
Enabling events on main process
Enabling events on 'copy'
Unhandled exception. System.InvalidOperationException: Cannot process request because the process (37344) has exited.
   at System.Diagnostics.Process.GetProcessHandle(Int32 access, Boolean throwIfExited)
   at System.Diagnostics.Process.GetOrOpenProcessHandle()
   at System.Diagnostics.Process.EnsureWatchingForExit()
   at System.Diagnostics.Process.set_EnableRaisingEvents(Boolean value)
   at Program.Main(String[] args) in C:\Users\mattkot\Desktop\temp\ConsoleApp1\Program.cs:line 35
```

This snippet is from a netcoreapp3.1 running on Windows 10 Version	10.0.18363 Build 18363

stephentoub replied:
I see. Thanks.  EnableRaisingEvents is eventually calling GetProcessHandle with throwIfExited:true.

For now, your workaround is fine.  Ideally though we'd avoid that exception in the first place rather than letting it be thrown and eating it.

MattKotsenas replied:
Thanks Stephen! To be very explicit about next steps, I see a few paths forward, which one do you prefer?

1. Merge PR as-is and file an issue about the `EnableRaisingEvents` behavior
2. Refactor the `EnableRaisingEvents` code path to avoid throwing the exception for this internal call-site, but keep the public behavior the same
3. Make a separate PR to change the behavior for `EnableRaisingEvents`, then rebase this PR on top of that one

My least favorite option is 2 because it hints that there's a flaw in the API and doesn't solve the issue for other devs. That said, the behavior of `EnableRaisingEvents` is a bit inconsistent either way depending on how you look at it. For instance, `MaxWorkingSet` and `Modules` _do_ throw if called after the process exits, so it's semi-reasonable to expect properties to throw on exited processes.

Let me know what path you prefer. Thanks!

stephentoub replied:
Eventually I think we should do 2. For now an issue is fine.

MattKotsenas replied:
Filed #1785.

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -62,6 +119,21 @@ public void CurrentProcess_WaitNeverCompletes(int milliseconds)
+            {
+                var token = cts.Token;
+                var process = Process.GetCurrentProcess();
+                var ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => process.WaitForExitAsync(token));
```
We do not use `var` if the right-hand side does is not a `new` statement. Makes review easier.

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,45 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+            p.Start();
+
+            // Verify we can try to wait for the process to exit multiple times
+            using (var cts = new CancellationTokenSource(0))
```
```suggestion
            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```

avoid numbers without an obvious meaning

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,45 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+            using (var cts = new CancellationTokenSource(0))
+            {
+                var token = cts.Token;
+                for (var i = 0; i < 2; i++)
```
```suggestion
                for (int i = 0; i < 2; i++)
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,45 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+                await p.WaitForExitAsync(cts.Token);
+                Assert.True(p.HasExited);
+            }
+            using (var cts = new CancellationTokenSource(0))
```
```suggestion
            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,45 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+            // before hand, in which case we're simply not testing exactly
+            // what we wanted to test, but everything should still work.
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            Task.Delay(10).ContinueWith(_ => p.Kill());
```
```suggestion
            _ = Task.Delay(10).ContinueWith(_ => p.Kill());
```

you can avoid the warnings using discard.

MattKotsenas replied:
Good to know! Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -107,6 +218,42 @@ public async Task SingleProcess_WaitAfterExited(bool addHandlerBeforeStart)
+            p.Kill();
+            Assert.True(await tcs.Task);
+
+            using (var cts = new CancellationTokenSource(0))
```
```suggestion
            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -143,6 +290,38 @@ public void SingleProcess_CopiesShareExitInformation()
+
+            Process[] copies = Enumerable.Range(0, 3).Select(_ => Process.GetProcessById(p.Id)).ToArray();
+
+            using (var cts = new CancellationTokenSource(0))
```
```suggestion
            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -143,6 +290,38 @@ public void SingleProcess_CopiesShareExitInformation()
+        [Fact]
+        public async Task SingleProcess_CopiesShareExitAsyncInformation()
+        {
+            Process p = CreateProcessLong();
```
```suggestion
            using Process p = CreateProcessLong();
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -143,6 +290,38 @@ public void SingleProcess_CopiesShareExitInformation()
+            using (var cts = new CancellationTokenSource(0))
+            {
+                var token = cts.Token;
+                var ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => p.WaitForExitAsync(token));
```
Avoid `var` without `new`.

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -143,6 +290,38 @@ public void SingleProcess_CopiesShareExitInformation()
+                Assert.True(p.HasExited);
+            }
+
+            using (var cts = new CancellationTokenSource(0))
```
```suggestion
            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -143,6 +290,38 @@ public void SingleProcess_CopiesShareExitInformation()
+                foreach (Process copy in copies)
+                {
+                    await copy.WaitForExitAsync(cts.Token);
+                    Assert.True(copy.HasExited);
```
This is a little confusing, may want to comment here as one might assume calling `WaitForExitAsync` with a cancelled token would throw.

MattKotsenas replied:
Added!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -169,6 +348,44 @@ public void WaitForPeerProcess()
+        [Fact]
+        public async Task WaitAsyncForPeerProcess()
+        {
+            Process child1 = CreateProcessLong();
```
```suggestion
            using Process child1 = CreateProcessLong();
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -169,6 +348,44 @@ public void WaitForPeerProcess()
+            Process child1 = CreateProcessLong();
+            child1.Start();
+
+            Process child2 = CreateProcess(async peerId =>
```
```suggestion
            using Process child2 = CreateProcess(async peerId =>
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -214,6 +431,55 @@ public void WaitForSignal()
+            Process p = CreateProcessPortable(RemotelyInvokable.WriteLineReadLine);
+            p.StartInfo.RedirectStandardInput = true;
+            p.StartInfo.RedirectStandardOutput = true;
+            var mre = new ManualResetEventSlim(false);
```
```suggestion
            using var mre = new ManualResetEventSlim(false);
```

MattKotsenas replied:
Fixed!

scalablecory commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -214,6 +431,55 @@ public void WaitForSignal()
+            const string successResponse = "Success";
+            const int timeout = 5 * 1000;
+
+            Process p = CreateProcessPortable(RemotelyInvokable.WriteLineReadLine);
```
```suggestion
            using Process p = CreateProcessPortable(RemotelyInvokable.WriteLineReadLine);
```

MattKotsenas replied:
Fixed!

MattKotsenas committed:
Follow style guidelines in ProcessWaitingTests

Per code review feedback, update the new unit tests in
ProcessWaitingTests to follow style guidelines.

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessTestBase.cs:
```diff
@@ -74,6 +86,20 @@ protected Process CreateProcess(Func<string, int> method, string arg, bool autoD
             return p;
         }
 
+        protected Process CreateProcess(Func<string, Task<int>> method, string arg, bool autoDispose = true)
```
It doesn't look like you use this autoDispose; you can just remove it and always call AddProcessForDispose.

MattKotsenas replied:
Since accepting the code review feedback in this thread this function isn't used anymore, so I just removed it.

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -62,6 +119,21 @@ public void CurrentProcess_WaitNeverCompletes(int milliseconds)
+        {
+            using (var cts = new CancellationTokenSource(milliseconds))
+            {
+                var token = cts.Token;
```
```suggestion
                CancellationToken token = cts.Token;
```

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -62,6 +119,21 @@ public void CurrentProcess_WaitNeverCompletes(int milliseconds)
+            using (var cts = new CancellationTokenSource(milliseconds))
+            {
+                var token = cts.Token;
+                var process = Process.GetCurrentProcess();
```
```suggestion
                Process process = Process.GetCurrentProcess();
```

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,43 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+            // Verify we can try to wait for the process to exit multiple times
+            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
+            {
+                var token = cts.Token;
```
```suggestion
                CancellationToken token = cts.Token;
```

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -143,6 +288,40 @@ public void SingleProcess_CopiesShareExitInformation()
+
+            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
+            {
+                var token = cts.Token;
```
```suggestion
                CancellationToken token = cts.Token;
```

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,43 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+            p.Start();
+
+            // Verify we can try to wait for the process to exit multiple times
+            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```
This is an expensive way of just doing `new CancellationToken(canceled: true)`.

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,94 @@ public bool WaitForExit(int milliseconds)
+        /// Calling this method will set <see cref="EnableRaisingEvents"/> to true.
+        /// </remarks>
+        public async Task WaitForExitAsync(CancellationToken cancellationToken = default)
+        {
```
There's enough expensive work happening as part of this, I suggest adding:
```C#
cancellationToken.ThrowIfCancellationRequested();
```

MattKotsenas replied:
Added!

MattKotsenas replied:
Added!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -62,6 +119,21 @@ public void CurrentProcess_WaitNeverCompletes(int milliseconds)
+            {
+                var token = cts.Token;
+                var process = Process.GetCurrentProcess();
+                OperationCanceledException ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => process.WaitForExitAsync(token));
```
You're passing in a token that's already been canceled by the time you call this.  As such, WaitForExitAsync should actually complete synchronously, and it'd be better to lock that behavior into the test, e.g.
```C#
Task t = process.WaitForExitAsync(token);
Assert.Equal(TaskStatus.Canceled, t.Status);
OperationCanceledException ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t);
```

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,43 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+                var token = cts.Token;
+                for (int i = 0; i < 2; i++)
+                {
+                    OperationCanceledException ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => p.WaitForExitAsync(token));
```
You're passing in a token that's already been canceled by the time you call this.  As such, WaitForExitAsync should actually complete synchronously, and it'd be better to lock that behavior into the test, e.g.
```C#
Task t = process.WaitForExitAsync(token);
Assert.Equal(TaskStatus.Canceled, t.Status);
OperationCanceledException ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t);
```

MattKotsenas replied:
Added!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,43 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
         }
 
+        [Fact]
+        public async Task SingleProcess_TryWaitAsyncMultipleTimesBeforeCompleting()
```
This isn't really doing what you want it do to, especially if you add the ThrowIfCancellationRequested I suggested earlier.  I think this shouldn't use a CancellationToken at all, e.g.
```C#
Process p = CreateProcessLong();
p.Start();

Task[] tasks = (from i in Enumerable.Range(0, 10) select p.WaitForExitAsync()).ToArray();
Assert.All(tasks, t => !t.IsCompleted);
p.Kill();
Task.WaitAll(tasks);
````

MattKotsenas replied:
I apologize, but I'm not following what it's doing versus should be doing. This test is basically an aync-ified clone of `SingleProcess_TryWaitMultipleTimesBeforeCompleting`.

Should the two tests diverge, or is there another test that should be added in addition?

stephentoub replied:
The token passed to WaitForExitAsync is already canceled prior to the call.  So it's going to get only to https://github.com/dotnet/runtime/pull/1278/files#diff-06c983bcaa17665febae824710b68dbeR1397 and never get any further.  If the goal is to test, for example, the event handle hookup and unhookup, this won't do that.  You would instead want to create a new CancellationTokenSource, pass its token to WaitForExitAsync, then cancel the token, and then await the previously returned task.

MattKotsenas replied:
Thanks for the additional info. I updated the unit test so that both cases are covered.

I also added comments to denote the different cases:
* Token is already canceled, so return synchronous
* Task is created, then token is canceled, so exercise cleanup code
* Task is awaited, then process is killed (the happy path)
* Waiting with a token on an already exited process, so return synchronous

Please let me know if there's anything else I missed. Thanks!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -82,6 +154,43 @@ public void SingleProcess_TryWaitMultipleTimesBeforeCompleting()
+                await p.WaitForExitAsync(cts.Token);
+                Assert.True(p.HasExited);
+            }
+            using (var cts = new CancellationTokenSource(millisecondsDelay: 0))
```
As noted earlier, this is just synchronously creating an already canceled token.

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -169,6 +348,44 @@ public void WaitForPeerProcess()
+            using Process child2 = CreateProcess(async peerId =>
+            {
+                Process peer = Process.GetProcessById(int.Parse(peerId));
+                Console.WriteLine("Signal");
```
Please remove the Console.WriteLines.

stephentoub replied:
Ah, nevermind, this is being redirected.

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -214,6 +431,55 @@ public void WaitForSignal()
+        public async Task WaitAsyncForSignal()
+        {
+            const string expectedSignal = "Signal";
+            const string successResponse = "Success";
```
Nit: consts should be PascalCased

MattKotsenas replied:
Fixed!

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -214,6 +431,55 @@ public void WaitForSignal()
+            Assert.Equal(1, linesReceived);
+
+            // Wait a little bit to make sure process didn't exit on itself
+            Thread.Sleep(100);
```
Let's make this just 1 instead of 100.

MattKotsenas replied:
Reduced to 1.

stephentoub commented on src/libraries/System.Diagnostics.Process/tests/ProcessWaitingTests.cs:
```diff
@@ -169,6 +348,44 @@ public void WaitForPeerProcess()
+                Console.WriteLine("Signal");
+                using (var cts = new CancellationTokenSource(WaitInMS))
+                {
+                    await peer.WaitForExitAsync(cts.Token);
```
I don't see any tests where WaitForExitAsync is used without a token on a process that exits successfully / normally.  Are there any?

MattKotsenas replied:
`SingleProcess_WaitAsyncAfterExited` waits without a token on a process that has already exited, but I hadn't written any tests that have a potentially unbounded wait in order to avoid possibly blocking CI.

Now that you've pointed it out though I've added another test `WaitAsyncForProcess` that doesn't pass a cancellation token, but still doesn't wait unboundedly.

MattKotsenas committed:
Update tests to follow coding guidelines

Per code review feedback, update tests to follow coding guidelines,
simplify creating cancellation tokens, and verify synchronous completion
of tasks in the exited / canceled case.

MattKotsenas committed:
Update WaitForExitAsync to early out in canceled case

Per code review feedback, add an early out for a non-exited process when
canceled.

MattKotsenas committed:
Add a test for completion without a cancellation token

Per code review feedback, add a test for a process that completes
normally and doesn't use a canellation token.

carlossanlop commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,100 @@ public bool WaitForExit(int milliseconds)
+        /// for the <paramref name="cancellationToken"/> to be canceled.
+        /// </summary>
+        /// <remarks>
+        /// Calling this method will set <see cref="EnableRaisingEvents"/> to true.
```
Thank you for documenting this new public API. Here's a small suggestion:
```suggestion
        /// Calling this method will set <see cref="EnableRaisingEvents"/> to <see langword="true" />.
```

MattKotsenas replied:
Thanks! Fixed!

carlossanlop commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,100 @@ public bool WaitForExit(int milliseconds)
+        /// <remarks>
+        /// Calling this method will set <see cref="EnableRaisingEvents"/> to true.
+        /// </remarks>
+        public async Task WaitForExitAsync(CancellationToken cancellationToken = default)
```
The triple slash comments are missing a `<returns></returns>` item. Can you please add it too?

MattKotsenas replied:
Added!

MattKotsenas committed:
Address PR feedback in xmldocs

Per code review feedback, update the xmldocs for `WaitForExitAsync` to
specify the language keyword for "true", and add a `<returns>` element.

MattKotsenas commented:
Friendly ping to reviewers:
 
* @carlossanlop
* @stephentoub
* @MihaZupan

Is there any additional follow up need here? Or is this OK to merge?

scalablecory commented:
This looks good to me now. @stephentoub anything else?

scalablecory commented:
@ericstj who is area owner here?

ericstj commented:
@krwq but he's out on leave.  I trust @stephentoub and @carlossanlop to review here.

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,103 @@ public bool WaitForExit(int milliseconds)
+        /// Calling this method will set <see cref="EnableRaisingEvents"/> to <see langword="true" />.
+        /// </remarks>
+        /// <returns>
+        /// A task that will complete when the process has exited.
```
Or when cancellation has been requested, or when something goes wrong.

MattKotsenas replied:
Updated!

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1340,6 +1341,103 @@ public bool WaitForExit(int milliseconds)
+            // CASE 3.2: PROCESS EXITS BEFORE REGISTERING HANDLER
+            // (See CASE 1.2)
+
+            if (!Associated) { throw new InvalidOperationException(SR.NoAssociatedProcess); }
```
```suggestion
            if (!Associated)
            {
                throw new InvalidOperationException(SR.NoAssociatedProcess);
            }
```

MattKotsenas replied:
Fixed!

MattKotsenas committed:
Update xmldocs per code review feedback

Per code review feedback, update xmldocs to list other conditions that
can cause the function to return.

MattKotsenas committed:
Update function guards per code review feedback

Per code review feedback, update the method guards to span multiple
lines to adhere to style guidelines.

MattKotsenas committed:
Refactor test to verify async as well as sync cancellation

Per code review feedback, update
SingleProcess_TryWaitAsyncMultipleTimesBeforeCompleting to test both the
case where the token is canceled before creating the Task as well as
after the task is already created.

MattKotsenas commented:
Friendly ping to reviewers:

* @stephentoub 

Is there any additional follow up need here? Or is this OK to merge?

stephentoub commented:
Thanks!
